pkg/ring/kv/multi.go

```diff
@@ -0,0 +1,341 @@
+	}
+	return true, nil
+}
+
+func (m *MultiClient) registerCancelFn(clientID int, fn context.CancelFunc) int {
+	m.inProgressMu.Lock()
+	defer m.inProgressMu.Unlock()
+
+	m.inProgressCnt++
+	id := m.inProgressCnt
+	m.inProgress[id] = clientInProgress{client: clientID, cancel: fn}
+	return id
+}
+
+func (m *MultiClient) unregisterCancelFn(id int) {
+	m.inProgressMu.Lock()
+	defer m.inProgressMu.Unlock()
+
+	delete(m.inProgress, id)
+}
+
+// Runs supplied fn with current primary client. If primary client changes, fn is restarted.
+// When fn finishes (with or without error), this method returns given error value.
+func (m *MultiClient) runWithPrimaryClient(origCtx context.Context, fn func(newCtx context.Context, primary kvclient) error) error {
+	cancelFn := context.CancelFunc(nil)
+	cancelFnID := 0
+
+	cleanup := func() {
+		if cancelFn != nil {
+			cancelFn()
+		}
+		if cancelFnID > 0 {
+			m.unregisterCancelFn(cancelFnID)
+		}
+	}
+
+	defer cleanup()
+
+	// This only loops if switchover to a new primary backend happens while calling 'fn', which is very rare.
+	for {
+		cleanup()
+		pid, kv := m.getPrimaryClient()
+
+		var cancelCtx context.Context
+		cancelCtx, cancelFn = context.WithCancel(origCtx)
+		cancelFnID = m.registerCancelFn(pid, cancelFn)
+
+		err := fn(cancelCtx, kv)
+
+		if err == nil {
+			return nil
+		}
+
+		if cancelCtx.Err() == context.Canceled && origCtx.Err() == nil {
+			// our context was cancelled, but outer context is not done yet. retry
+			continue
+		}
+
+		return err
+	}
+}
+
+// Get is a part of kv.Client interface
+func (m *MultiClient) Get(ctx context.Context, key string) (interface{}, error) {
+	var result interface{}
+	err := m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		var err2 error
+		result, err2 = primary.client.Get(newCtx, key)
+		return err2
+	})
+	return result, err
+}
+
+// CAS is a part of kv.Client interface
+func (m *MultiClient) CAS(ctx context.Context, key string, f func(in interface{}) (out interface{}, retry bool, err error)) error {
+	return m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		return primary.client.CAS(newCtx, key, f)
+	})
+}
+
+// WatchKey is a part of kv.Client interface
+func (m *MultiClient) WatchKey(ctx context.Context, key string, f func(interface{}) bool) {
+	_ = m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		primary.client.WatchKey(newCtx, key, f)
+		return newCtx.Err()
+	})
+}
+
+// WatchPrefix is a part of kv.Client interface
+func (m *MultiClient) WatchPrefix(ctx context.Context, prefix string, f func(string, interface{}) bool) {
+	_ = m.runWithPrimaryClient(ctx, func(newCtx context.Context, primary kvclient) error {
+		primary.client.WatchPrefix(newCtx, prefix, f)
+		return newCtx.Err()
+	})
+}
+
+// watchChanges performs mirroring -- it watches changes in the primary client, and puts them into secondary.
+func (m *MultiClient) watchChanges(prefix string) {
```

Comment:
I think it would achieve the same, at the cost of making CAS slower for clients. But that is probably just fine. (There is question of what to do when CAS to primary succeeds, but CAS to secondary fails, but the only reasonable thing we can do is to ignore it... we should not call `f` again)